akka-http/src/main/scala/akka/http/server/directives/CodingDirectives.scala

```diff
@@ -16,46 +18,64 @@ trait CodingDirectives {
   import MiscDirectives._
   import RouteDirectives._
   import CodingDirectives._
+  import HeaderDirectives._
 
   // encoding
 
   /**
-   * Wraps its inner Route with encoding support using the given Encoder.
-   */
-  def encodeResponse(encoder: Encoder): Directive0 =
-    responseEncodingAccepted(encoder.encoding) &
-      mapResponse(encoder.encode(_)) &
-      cancelRejections(classOf[UnacceptedResponseEncodingRejection])
-
-  /**
    * Rejects the request with an UnacceptedResponseEncodingRejection
    * if the given encoding is not accepted for the response.
    */
   def responseEncodingAccepted(encoding: HttpEncoding): Directive0 =
     extract(_.request.isEncodingAccepted(encoding))
-      .flatMap(if (_) pass else reject(UnacceptedResponseEncodingRejection(encoding)))
+      .flatMap(if (_) pass else reject(UnacceptedResponseEncodingRejection(Set(encoding))))
 
   /**
-   * Wraps its inner Route with response compression, using the specified
-   * encoders in the given order of preference.
-   * If no encoders are specifically given Gzip, Deflate and NoEncoding
-   * are used in this order, depending on what the client accepts.
+   * Wraps its inner Route with response encoding, using one of the specified
+   * encoders negotiated with the encodings the client accepts.
+   *
+   * The negotiation logic is defined by http://tools.ietf.org/html/rfc7231#section-5.3.4.
    */
-  def compressResponse(encoders: Encoder*): Directive0 =
-    theseOrDefault(encoders).map(encodeResponse(_)).reduce(_ | _)
+
+  def encodeResponseWith(encoders: Encoder*): Directive0 = _encodeResponse(immutable.Seq(encoders: _*))
 
   /**
-   * Wraps its inner Route with response compression if and only if the client
-   * specifically requests compression with an `Accept-Encoding` header.
+   * Wraps its inner Route with response compression unless the client denied compression with an
+   * `Accept-Encoding` header.
    */
-  def compressResponseIfRequested(): Directive0 = compressResponse(NoCoding, Gzip, Deflate)
+  def encodeResponse: Directive0 = _encodeResponse(DefaultCoders)
```

Comment:
Having thought once more about this I wonder if we should just get rid of this directive and stick to explicitly supplied encoders with a note that their ordering only influences the no-` Accept-Encoding`-header-supplied case and otherwise doesn't influence encoding negotiation.